Ease recording walls and tubes

Brett208 opened the pull request:
Closes #11

This needs actual testing. Just wanted to submit for review before I start implementing in actual test cases in mission code.

Brett208 committed:
Create a LOCATION swap routine

This is to replace the swap routine within CreateTubeOrWallLine

Brett208 committed:
Move CreateTubeOrWallLine into .cpp file

This function is becoming 'private' and will be wrapped by 2 public interface functions CreateTube and CreateWall

Brett208 committed:
Make CreateTubeOrWallLine use LOCATION instead of separate ints for x and y

Brett208 committed:
Add CreateTubeLine and CreateWallLine functions

Default CreateWallLine to a map_id::mapWall

Brett208 committed:
Add RecordTubeLine and RecordWallLine functions

Brett208 committed:
Update BaseBuilder to use new CreateTubeLine and CreateWallLine routines

DanRStevens commented on OP2Helper.h:
```diff
@@ -52,9 +52,6 @@ bool operator!= (const LOCATION& loc1, const LOCATION &loc2);
 void InitPlayerResources(int playerNum, const ResourceSet& resourceSet = CES1ResourceSet);
 
-// Constructing Tubes/Walls
-void CreateTubeOrWallLine(int x1, int y1, int x2, int y2, map_id type);
```
Rather than removing public access, and breaking old code, I was thinking maybe it should be marked `[[deprecated]]` instead.

I see you've later changed the parameters from unpacked `int`s to `LOCATION` structs, so that would break old code. Not sure if I want to force people to pack parameters into `LOCATION` structs, though I certainly don't object to allowing it as a convenience.

DanRStevens commented on OP2Helper.h:
```diff
@@ -52,6 +52,16 @@ bool operator!= (const LOCATION& loc1, const LOCATION &loc2);
+
+// Creates walls in a straight line or L pattern
+// L patterns draw counter - clockwise between the two points
+void CreateWallLine(LOCATION loc1, LOCATION loc2, map_id wallType = map_id::mapWall);
```
I think if you're going to split "Tube" from wall, you probably need to go all the way and split each wall type from each other.

DanRStevens commented on OP2Helper.cpp:
```diff
@@ -78,6 +78,39 @@ void CreateTubeOrWallLine(LOCATION loc1, LOCATION loc2, map_id type)
+	}
+}
+
+void RecordWallLine(BuildingGroup& buildingGroup, LOCATION loc1, LOCATION loc2, map_id wallType)
```
Hmm, that's quite noticeable duplication here. I see there is a distinction between `RecordTube` and `RecordWall`, so that's a bit annoying to work around. Maybe with templates or lambdas, the difference can be worked around. Might even be able to share code with the `Create` methods.

On another note, I can see why you split between "Tube" and "Wall". I think I still like the idea of unpacking the wall types further though.

DanRStevens reviewed (commented):
Well that explains a bit that perhaps I didn't fully realize before. Looks to be on the right track. I'd probably try to make a few adjustments that could maintain backwards compatibility. I see there are some code duplication challenges. I'm actually not quite certain off the top of my head how to overcome the source code duplication, though I'm sure it can be done. There's probably an answer within lambdas or `std::function`.

Brett208 commented:
I'm going to try and summarize @DanRStevens review:

1. Restore the original CreateTubeOrWall function and mark as deprecated (I can transfer the int x and y values into LOCATION structs within CreateTubeOrWall to use the new Swap LOCATION function)
1. Break out 4 new public functions: CreateMicrobeWallLine, CreateLavaWallLine, RecordMicrobeWallLine, RecordLavaWallLine. (I'm on the fence about this. It seemed more natural to pass a parameter for the wall type to me. I will change though since it seems preferred by @DanRStevens)
1. Consider attempting to reduce code duplication. (I typically find functors, lambdas, and delegates more difficult to read. I will take a brief look at using lambdas to solve. If not easy, would prefer to proceed with the duplication)

DanRStevens commented:
Sounds like a decent plan. A few thoughts on possible options:
1. It might be easier to unpack structs, rather than pack structs. Though I may be thinking more at the assembly level that source language level.
2. If you want to avoid splitting out too many functions, my alternative suggestion was using a more inclusive name, such as `CreateEarthworkerItem`. For a corresponding `RecordEarthworkerItem`, it could internally decided if it should call `RecordTube` or `RecordWall`. (Does it even matter which one you call? I imagine both record methods are the same internally).
3. I can possibly help with removing code duplication later on. It's a curious problem.

Brett208 committed:
Use lambda functions to remove code duplication

Brett208 committed:
Add CreateWallOrTubeLine back in as a deprecated function

Brett208 commented on OP2Helper.cpp:
```diff
@@ -18,6 +19,13 @@ const ResourceSet MultiResourceSet = {
 };
 
+// Perform recordFunction discretely along a line or L shape
+void ExecuteAcrossLine(LOCATION loc1, LOCATION loc2, const std::function <void(LOCATION&)>& recordFunction);
```
passing LOCATION& here as an argument into the lambda would be more appropriate as const LOCATION&, but underlying Outpost2DLL does not support.

Brett208 commented on OP2Helper.h:
```diff
@@ -52,9 +52,25 @@ bool operator!= (const LOCATION& loc1, const LOCATION &loc2);
+void RecordMicrobeWallLine(BuildingGroup& buildingGroup, const LOCATION& loc1, const LOCATION& loc2);
+
+[[deprecated("CreateTubeOrWallLine is deprecated. Use CreateWallLine or CreateTubeLine instead.")]]
 void CreateTubeOrWallLine(int x1, int y1, int x2, int y2, map_id type);
```
This function is instantiated so that it can be wholesale deleted in a future version without affecting any other function/code within OP2Helper.

DanRStevens replied:
Really minor point: For technical writing we should probably try to list things in the same order whenever possible, hence "TubeOrWall" -> "Tube", "Wall".

Might also be good to indicate there are multiple wall types. Perhaps:
```cpp
[[deprecated("CreateTubeOrWallLine is deprecated. Use CreateTubeLine or Create[|Lava|Microbe]WallLine instead.")]]
```
Alternatively maybe `Create<X>WallLine` might be another syntax. That might look a bit too messy though. I'm not overly against leaving it simply as `CreateWallLine`.

Brett208 commented on OP2Helper.cpp:
```diff
@@ -38,6 +46,89 @@ void InitPlayerResources(int playerNum, const ResourceSet& resourceSet)
+	);
+}
+
+void ExecuteAcrossLine(LOCATION loc1, LOCATION loc2, const std::function <void(LOCATION&)>& recordFunction)
```
Happy to entertain different name for function and passed in lambda. Don't want to spend forever trying to find perfect name either though.

DanRStevens replied:
I would name it just `function`. The function is being used for both Create and Record, so probably best to drop the "record" part as it is too specific.

Brett208 replied:
I meant record in the sense of generally record an event at each location in the L. However, it is confusing sense there is a record wall/tube as well. I think function as the name will work fine. I assumed it was a keyword for some reason, but looks like you added it fine.

Brett208 reviewed (commented):
Apparently my knowledge of C# lamdbas paid good dividends since this wasn't too hard to implement. The code is now a lot harder for someone with limited lambda programming to understand, although after putting the effort in I'd like to see it included.

I haven't checked the code out yet in maps, which should be done before merging.

Thanks,
Brett

DanRStevens reviewed (commented):
Interesting note about the `LOCATION&` parameter. 

You're right, the lack of `const` can be suspicious, though I often just assume the author either missed the `const`, or wasn't aware enough of const-correctness to know to put it in. After all, so much previous code used `char*` for strings, rather than `const char*`.

In terms of the `ExecuteAcrossLine`'s function parameter declaration, I suppose if you were passing independent (x, y) values, the lack of `const` detail could be hidden internally. Whether or not a local variable in the lambda method gets modified after use is of little consequence to the caller. The values are already unpacked in `ExecuteAcrossLine`, so it would be quite natural to pass unpacked values. On the receiving side, the Create methods use unpacked (x, y) values natively. In contrast, the Record methods take a packed `LOCATION&`, so would need a local variable to be constructed. This would need to be an actual variable, as non-const references do not bind to rvalues, so a temporary inline construction would not work. So, slightly messy with the variable declaration, but not too bad. Such a setup would reduce the packing/unpacking for the Create set of methods, and be the same for the Record set of methods.

Actually, it looks like `ExecuteAcrossLine` is passing a temporarily constructed `LOCATION` value to the `recordFunction`, even though the parameter is declared to accept `LOCATION&`. I think you may be using a Microsoft specific C++ language extension there. Yes, actually, Mingw complains about it:
```cpp
OP2Helper.cpp: In function ‘void ExecuteAcrossLine(LOCATION, LOCATION, const std::function<void(LOCATION&)>&)’:
OP2Helper.cpp:114:40: error: no match for call to ‘(const std::function<void(LOCATION&)>) (LOCATION)’
   recordFunction(LOCATION(x, horizEdge));
```

DanRStevens committed:
Use unpacked parameters for Tube/Wall line

This avoids extra packing/unpacking for the Create methods, while
working basically the same for the Record methods.

This avoids Microsoft specific extensions to the C++ language, which
allowed non-const lvalues to bind to rvalue temporaries. Instead, the
variable creation is made explicit.

DanRStevens committed:
Rename `recordFunction` to just `function`

Brett208 committed:
Merge pull request #13 from OutpostUniverse/useUnpackedParameters

Use unpacked parameters

Brett208 committed:
Improve deprecated function warning

Brett208 commented:
Thanks for catching the Microsoft specific language extension. I've used that several times before. It seems so intuitive, I wonder why the standard hasn't incorporated it. I thought your branch changes were fine.

-Brett

DanRStevens reviewed (approved):
I think standard C++ disallows that for the same reason you get a warning from:
```cpp
x + 1;
```

It's a perfectly valid calculation, but the result is thrown away without ever being used. Same thing when you write to a temporary variable with no surviving references and which is right about to be destroyed. Unless you're explicit about allowing it (rvalue references), the standard doesn't allow it to happen by default.

The other part of it might be to remind people to use `const` on parameters like that which don't return values. :wink:

Brett208 commented:
Tested the new functions in a mission. Everything worked correctly except the AI didn't build the Microbe and Lava walls. I assume because they didn't have the tech researched.
